### ISSUE TITLE
Data dictionary for each possible parameter to the model

### ISSUE THREAD

**Issue by arik-shurygin** [CDCgov/DynODE#291]
breakdown each accepted parameter for each configuration file (global, initializer, inferer). Make sure to include the definition, the accepted shape, what each dimension in the shape implies, and possibly downstream parameters that are automatically generated when that parameter is provided.

**Comment by edbaskerville** [comment]
I'd love for everything about a parameter to be in the same place (currently, `config.py`):

* Name (identifier in code)
* Human-readable description
* Type
* If array, shape in terms of other parameters

The descriptions could be then used to generate model documentation, and the shapes could be used for automatic validation.

e.g.

```
{
    "name": "WANING_PROTECTIONS",
    "description": "Immune protection for each waning compartment as a fraction of maximum protection."
    "type": np.array,
    "shape": lambda p: (p.NUM_WANING_COMPARTMENTS",),
},
```

You can probably do this with manual validations in the current system?

Of course I'd prefer that this all be part of a real type system, but hey, Python

**Comment by arik-shurygin** [comment]
> I'd love for everything about a parameter to be in the same place (currently, `config.py`):
> 
> * Name (identifier in code)
> * Human-readable description
> * Type
> * If array, shape in terms of other parameters
> 
> The descriptions could be then used to generate model documentation, and the shapes could be used for automatic validation.
> 
> e.g.
> 
> ```
> {
>     "name": "WANING_PROTECTIONS",
>     "description": "Immune protection for each waning compartment as a fraction of maximum protection."
>     "type": np.array,
>     "shape": lambda p: (p.NUM_WANING_COMPARTMENTS",),
> },
> ```
> 
> You can probably do this with manual validations in the current system?
> 
> Of course I'd prefer that this all be part of a real type system, but hey, Python

This is a great idea, it also fits in with the efforts outlines in #180, it is likely worth it to specific attributes to the config class with pydoc comments, allowing api generation easily when #217 is tackled

**Comment by edbaskerville** [comment]
> This is a great idea, it also fits in with the efforts outlines in #180, it is likely worth it to specific attributes to the config class with pydoc comments, allowing api generation easily when #217 is tackled

Meaning make parameters be attributes of an object? Or something else?

A runtime type-ier version of this could be...

```python
from dynode.parameters import parameter, closed_interval

class ScenariosModelParameters(ModelParameters):
    waning_protections: np.array = parameter(
        description = "blah blah blah",
        shape = lambda p: (p.num_waning_compartments,),
        range: closed_interval(0.0, 1.0),
        validation = lambda p: some_extra_validation(),
    )
```

I'm remembering my SQLAlchemy ORM here: https://docs.sqlalchemy.org/en/20/orm/quickstart.html

**Comment by arik-shurygin** [comment]
Yeah something like this, this way when someone refers to a Config object they are able to get type hints and information about each parameter they can reference from it.

Currently the code that adds attributes to `Config` looks like:
https://github.com/CDCgov/DynODE/blob/d2fb41fac338ded891ac27b221a7b95b2fb71f9f/src/resp_ode/config.py#L33-L57

specifically 
https://github.com/CDCgov/DynODE/blob/d2fb41fac338ded891ac27b221a7b95b2fb71f9f/src/resp_ode/config.py#L54

is the main point in which all the keys in the JSON are added to the config. If we instead call a bunch of `settrs` and define ahead of time all of the different types of parameters that come by default in the `Config` class. This way we get code autocomplete and other quality of life features.

**Comment by arik-shurygin** [comment]
@edbaskerville and I had a separate conversation about using Chex as a parameters library since we are already using Jax in DynODE as a whole, leaving the link [here](https://github.com/[google-deepmind/chex](https://github.com/google-deepmind/chex)) for when this issue is taken on

**Comment by edbaskerville** [comment]
Just to clarify, the general idea is to use dataclasses as a convenient built-in notation for typed model parameters. chex implements a jax-compatible dataclass, but we wouldn't necessarily need to use it if it proves cumbersome--can always transform to something else at runtime.

**Comment by arik-shurygin** [comment]
Closing as `pydantic` configs now allow for this human readability within all classes of the `dynode.model_configuration` module in the development branch.
